Move docker env var settings handling out of bash

rjernst opened the pull request:
In docker ES allows settings to be set via environment variables. This
is currently handled in complex bash logic. This commit moves that logic
into EnvironmentAwareCommand where the rest of the initial settings are
found.

relates #85758

rjernst committed:
Move docker env var settings handling out of bash

In docker ES allows settings to be set via environment variables. This
is currently handled in complex bash logic. This commit moves that logic
into EnvironmentAwareCommand where the rest of the initial settings are
found.

relates #85758

elasticmachine commented:
Pinging @elastic/es-core-infra (Team:Core/Infra)

rjernst committed:
spotless

rjernst committed:
debug

rjernst committed:
Merge branch 'master' into cli/env_passthrough

rjernst committed:
more debug

rjernst committed:
remove old code

pugnascotia committed:
Fix xpack security detection for Docker tests

Previously, in order to determine whether security had been disabled via
environment variables in Docker, we were able to check the command line
switches to the Elasticsearch process. That doesn't work now because
Elasticsearch processes the environment variables itself. Therefore,
change the check to look at the environment instead.

pugnascotia committed:
Imports

pugnascotia commented on qa/os/src/test/java/org/elasticsearch/packaging/util/docker/Docker.java:
```diff
@@ -78,7 +78,7 @@ public class Docker {
      * the same name.
      */
-    static String containerId = null;
+    public static String containerId = null;
```
I don't think we need this now?

pugnascotia commented on server/src/test/java/org/elasticsearch/common/cli/EnvironmentAwareCommandTests.java:
```diff
@@ -0,0 +1,111 @@
+        };
+    }
+
+    public void testNonDockerEnvVarsIgnored() throws Exception {
```
```suggestion
    /**
     * Check that env vars are not respected when the build type is not Docker.
     */
    public void testNonDockerBuildIgnoresEnvVars() throws Exception {
```

pugnascotia commented on server/src/test/java/org/elasticsearch/common/cli/EnvironmentAwareCommandTests.java:
```diff
@@ -0,0 +1,111 @@
+        execute();
+    }
+
+    public void testDockerEnvSettings() throws Exception {
```
```suggestion
    /**
     * Check that when the build type is Docker, then the different supported styles of passing settings
     * via env vars are handled.
     */
    public void testDockerBuildRespectsEnvVars() throws Exception {
```

pugnascotia commented on server/src/test/java/org/elasticsearch/common/cli/EnvironmentAwareCommandTests.java:
```diff
@@ -0,0 +1,111 @@
+        mockEnvVars.put("setting.Ignored", "baz");
+        callback = env -> {
+            Settings settings = env.settings();
+            assertThat(settings.hasValue("simple.setting"), is(true));
```
What do you think about also checking that the settings have the correct values as well?

pugnascotia commented on server/src/test/java/org/elasticsearch/common/cli/EnvironmentAwareCommandTests.java:
```diff
@@ -0,0 +1,111 @@
+        execute();
+    }
+
+    public void testDockerEnvSettingsOverride() throws Exception {
+        // docker env takes precedence over settings on the command line
```
```suggestion
    /**
     * Check that with the Docker build, env vars takes precedence over settings on the command line.
     */
    public void testDockerBuildEnvVarsOverrideCommandLine() throws Exception {
```

pugnascotia commented on server/src/test/java/org/elasticsearch/common/cli/EnvironmentAwareCommandTests.java:
```diff
@@ -0,0 +1,111 @@
+import static org.hamcrest.Matchers.equalTo;
+import static org.hamcrest.Matchers.is;
+
+public class EnvironmentAwareCommandTests extends CommandTestCase {
```
Can we also please replicate `DockerTests.test087EnvironmentVariablesInIncorrectFormatAreIgnored()`? It checks that env vars, which look like maybe they ought to be respected, are ignored due to being malformed.

You'll also need to remove these tests from `DockerTests`, because they rely on being able to inspect the Elasticsearch command line for options:

   * `test086EnvironmentVariablesInSnakeCaseAreTranslated`
   * `test087EnvironmentVariablesInIncorrectFormatAreIgnored`
   * `test087EnvironmentVariablesInIncorrectFormatAreIgnored`

pugnascotia reviewed (changes requested):
There are a number of debug statements that need to be removed as well.

pugnascotia committed:
Formatting

rjernst committed:
cleanup debug

rjernst committed:
cleanup tests and add some comments

rjernst commented:
Thanks Rory! I think I have a addressed all your comments. I normally don't like javadoc comments on test methods *and* having super descriptive names (which are actually harder to read IMO becauseOfTheLongRunningCamelCaseThatIsHardToVisuallyParse ;) But I added some simple comments to each method to describe what they are checking and standardized on method naming.

rjernst committed:
spotless

pugnascotia committed:
Fix ServerUtils.waitForElasticsearch

We were checked whether security was enabled or disabled in the
environment by running the `env` command and passing it through `grep`.
Unforunately, and somewhat obviously, `grep` exits non-zero if it
doesn't find a match, which causes `Shell.run` to throw and exception.
We could fix it with `runScriptIgnoreExitCode` instead of `run`, but
instead I chose just fetch all the `env` values and check them in the
Java code.

pugnascotia committed:
Stupid formatting change

pugnascotia committed:
Merge remote-tracking branch 'upstream/master' into cli/env_passthrough

rjernst commented:
Thanks for the docker test fixes Rory. I think this is ready for a final review now that all tests are passing.

rjernst committed:
Merge branch 'master' into cli/env_passthrough
